// red9resource/test/stub.resource.description.js
var _ = require('underscore')._;
var VError = require('verror');

/**
 * Based on the event resource
 *
 */

module.exports = {
    name: 'event',
    tableName: 'event',

    mapping: [
        {
            cassandraKey: 'time',
            cassandraType: 'timestamp',
            jsKey: 'time',
            jsType: 'timestamp'
        }  ,
        {
            cassandraKey: 'type',
            cassandraType: 'varchar',
            jsKey: 'type',
            jsType: 'string'
        },
        {
            cassandraKey: 'id',
            cassandraType: 'uuid',
            jsKey: 'id',
            jsType: 'string'
        }
    ],

    checkResource: function (event, callback) {
        if (event.time < 0) {
            callback(new VError('time %s is greater than or equal to event 0', event.time));
            return;
        }
        callback(null, event);
    },

    populateDefaults: function (newEvent) {
    },
    populateOnCreate: function (newEvent) {
        // Fill in the extras
    }
};
