// red9resource/lib/cassandra.js
var _ = require('underscore')._;
var cassandra;
var cassandraDriver = require('cassandra-driver');
var async = require('async');
var Boom = require('boom');

exports.init = function (config, callback) {
    cassandra = new cassandraDriver.Client({
        contactPoints: config.cassandraHosts,
        keyspace: config.cassandraKeyspace,
        authProvider: new cassandraDriver.auth.PlainTextAuthProvider(config.cassandraUsername, config.cassandraPassword)
    });

    cassandra.connect(function (err, result) {
        if (_.isFunction(callback)) {
            callback(err);
        } else if (err) {
            console.log('Error: ' + err);
        }
    });

    cassandra.on('log', function (level, className, message, furtherInfo) {
        //console.log('log event: %s -- %s', level, message);
    });
};

/**
 * There's no configuration check here since that's a programmer error, and should be caught very quickly.
 *
 * There's no configuration check here since that's a programmer error, and should be caught very quickly.
 *
 * rowCallback (mappedResource, currentRow)
 * callback (err, rowCount)
 * @param options
 * - query
 * - rowCallback
 * - callback
 * - parameters
 * - hints
 */
exports.execute = function (options) {
    var query = options.query;
    var rowCallback = options.rowCallback;
    var callback = options.callback;

    var parameters = options.parameters;
    if (!_.isArray(parameters)) {
        parameters = [];
    }

    console.log('Query: ' + query);
    console.dir(parameters);
    console.dir(options.hints);
    var queryOptions = {
        autoPage: true,
        prepare: false,
        fetchSize: 250
    };

    if (_.isArray(options.hints)) {
        queryOptions.hints = options.hints;
    }

    cassandra.eachRow(query, parameters,
        queryOptions,
        function (n, row) {
            if (_.isFunction(rowCallback)) {
                rowCallback(row, n);
            }
        },
        function (err, rowCount) {
            if (err) {
                callback(Boom.wrap(err, 500, 'Internal cassandra error. Please report.'));
            } else {
                callback(null, rowCount);
            }
        });
};

exports.getDatasetCount = function (datasetId, doneCallback) {
    var queries = [
        {
            query: 'SELECT count(*) FROM event WHERE dataset = ?',
            key: 'event'
        },
        {
            query: 'SELECT count(*) FROM comment WHERE resource = ?',
            key: 'comment'

        },
        {
            query: 'SELECT count(*) FROM video WHERE dataset = ?',
            key: 'video'
        }
    ];

    var queryOptions = {
        prepare: true,
        hints: ['uuid']
    };

    async.reduce(queries, {},
        function (memo, query, callback) {
            cassandra.execute(query.query, [datasetId], queryOptions, function (err, result) {
                if (err) {
                    console.log('Cassandra Error: ' + err);
                }
                memo[query.key] = result.rows[0].count.low;
                callback(null, memo);
            });
        },
        function (err, result) {
            if (err) {
                console.log('Error: ' + err);
            }
            doneCallback(null, result);
        });
};
