// red9resource/lib/resource.description.user.js
"use strict";

var Joi = require('joi');
var validators = require('./validators');

var basicModel = {
    id: validators.id,
    email: Joi.string().email().description('user email address. Used as "second" primary key for authentication'),
    displayName: Joi.string().description("user's preferred public handle"),
    givenName: Joi.string().description('first name'),
    familyName: Joi.string().description('last name'),
    preferredLayout: Joi.object().options({className: 'preferredLayout'}).default({}).description('layouts for user'),
    picture: Joi.string().description('link to a profile picture.'),
    gender: Joi.string().valid('male', 'female', 'unknown').description('gender')
    //affiliations: Joi.array().includes(Joi.string()).description('organization affiliations of this user'),
};

var resourceName = 'user';


module.exports = {
    name: resourceName,
    tableName: 'user',

    models: {
        model: basicModel,
        create: Joi.object({
            email: basicModel.email.required(),
            displayName: basicModel.displayName.required(),
            givenName: basicModel.givenName.required(),
            familyName: basicModel.familyName.required(),
            preferredLayout: basicModel.preferredLayout
        }).options({className: resourceName + '.create'}),
        update: Joi.object({
            email: basicModel.email,
            displayName: basicModel.displayName,
            givenName: basicModel.givenName,
            familyName: basicModel.familyName,
            preferredLayout: basicModel.preferredLayout
        }).options({className: resourceName + '.update'}),
        resultModel: Joi.object({
            id: basicModel.id.required(),
            email: basicModel.email.required(),
            displayName: basicModel.displayName.required(),
            givenName: basicModel.givenName.required(),
            familyName: basicModel.familyName.required(),
            preferredLayout: basicModel.preferredLayout,
            picture: basicModel.picture.required(),
            gender: basicModel.gender.required()
        }).options({className: resourceName}),
        search: {
            id: validators.idCSV,
            idList: validators.idCSV,
            email: basicModel.email,
            displayName: basicModel.displayName
        }
    },

    mapping: [
        {
            cassandraKey: 'id',
            cassandraType: 'uuid',
            jsKey: 'id',
            jsType: 'string'
        },
        {
            cassandraKey: 'display_name',
            cassandraType: 'varchar',
            jsKey: 'displayName',
            jsType: 'string'
        },
        {
            cassandraKey: 'first',
            cassandraType: 'varchar',
            jsKey: 'givenName',
            jsType: 'string'
        },
        {
            cassandraKey: 'last',
            cassandraType: 'varchar',
            jsKey: 'familyName',
            jsType: 'string'
        },
        {
            cassandraKey: 'preferred_layout',
            cassandraType: 'varchar',
            jsKey: 'preferredLayout',
            jsType: 'object'
        },
        {
            cassandraKey: 'email',
            cassandraType: 'varchar',
            jsKey: 'email',
            jsType: 'string'
        },
        {
            cassandraKey: 'picture',
            cassandraType: 'varchar',
            jsKey: 'picture',
            jsType: 'string'
        },
        {
            cassandraKey: 'gender',
            cassandraType: 'varchar',
            jsKey: 'gender',
            jsType: 'string'
        }
    ],

    checkResource: function (user, callback) {
        // TODO(SRLM): Add some checks here:
        // - email does not exist

        callback(null, user);
    },
    expand: function (parameters, user, callback) {
        callback(null, user);
    }
};
