// red9resource/lib/resource.description.user.js
var _ = require('underscore')._;

module.exports = {
    name: 'user',
    tableName: 'user',



    mapping: [
        {
            cassandraKey: 'id',
            cassandraType: 'uuid',
            jsKey: 'id',
            jsType: 'string'
        },
        {
            cassandraKey: 'display_name',
            cassandraType: 'varchar',
            jsKey: 'displayName',
            jsType: 'string'
        },
        {
            cassandraKey: 'first',
            cassandraType: 'varchar',
            jsKey: 'givenName',
            jsType: 'string'
        },
        {
            cassandraKey: 'last',
            cassandraType: 'varchar',
            jsKey: 'familyName',
            jsType: 'string'
        },
        {
            cassandraKey: 'preferred_layout',
            cassandraType: 'varchar',
            jsKey: 'preferredLayout',
            jsType: 'object'
        },
        {
            cassandraKey: 'email',
            cassandraType: 'varchar',
            jsKey: 'email',
            jsType: 'string'
        }
    ],

    checkResource: function (user, callback) {
        // TODO(SRLM): Add some checks here:
        // - email is valid
        // - name is valid (no odd characters)
        // - email does not exist
        // - make sure preferred layout is valid

        callback(null);
    },
    expand: function (parameters, user, callback) {
        callback(null, user);
    },
    populateDefaults: function (newUser) {

    },
    populateOnCreate: function (newUser) {
        // TODO(SRLM): Add in a better preferred layout here.
        newUser.preferredLayout = {};
    }
};
