// red9resource/lib/resource.js
/** Master file to provide access to resources
 *
 *
 */
var _ = require('underscore')._;

var crud = require('./resource.crud.js');

var eventDescription = require('./resource.description.event');
var userDescription = require('./resource.description.user');
var commentDescription = require('./resource.description.comment');
var videoDescription = require('./resource.description.video');
var layoutDescription = require('./resource.description.layout');
var datasetDescription = require('./resource.description.dataset');

/**
 *
 * @param config With the following keys:
 * - cassandraHosts (array of IP/URL:PORTs),
 * - cassandraKeyspace (string)
 * @param callback {err}
 *
 */
exports.init = function (config, callback) {
    require('./cassandra').init(config, callback);
};

/**
 *
 * @param resourceDescription
 * @returns {{}}
 */
function addResource(resourceDescription) {
    var result = {};

    result.name = resourceDescription.name;
    result.models = resourceDescription.models;

    /**
     *
     * @param newResource
     * @param callback {err, ???}
     * @param deepMigrate {bool} never set to true.
     */
    result.create = function (newResource, callback, deepMigrate) {
        crud.create(resourceDescription, newResource, callback, deepMigrate);
    };

    /**
     *
     * @param query
     * @param options
     * @param rowCallback
     * @param callback (err, totalRows)
     */
    result.find = function (query, options, rowCallback, callback) {
        crud.find(resourceDescription, query, options, rowCallback, callback);
    };

    /**
     *
     * @param id
     * @param updatedEvent can be entire event (non-editable keys will be ignored). Must have at least one editable key
     * @param callback {function} {err, updatedResource}
     */

    result.update = function (id, updatedEvent, callback) {
        crud.update(result.find, resourceDescription, id, updatedEvent, callback);
    };

    /** DELETE from database
     *
     * Possible errors?
     *
     * If the resource doesn't exist we get something like this:
     * { [VError: event 678ad3be-9d48-40e1-c217-db866f9972a9 does not exist (0 results found)]
  jse_shortmsg: 'event 678ad3be-9d48-40e1-c217-db866f9972a9 does not exist (0 results found)',
  jse_summary: 'event 678ad3be-9d48-40e1-c217-db866f9972a9 does not exist (0 results found)',
  message: 'event 678ad3be-9d48-40e1-c217-db866f9972a9 does not exist (0 results found)' }
     *
     * Obviously, this should be made more machine parsable so that the caller can figure out if
     * there's a real error (no DB, etc.) or the resource just doesn't exist.
     *
     *
     *
     * @param id
     * @param callback {function} (err, deletedResource)
     */
    result.delete = function (id, callback) {
        crud.delete(result.find, resourceDescription, id, callback);
    };


    // Give resourceDescription access to other resources
    //resourceDescription.resource = exports;
    if (_.isFunction(resourceDescription.setResources)) {
        resourceDescription.setResources(exports);
    }

    // Add r to the list of resources
    return result;
}


exports[eventDescription.name] = addResource(eventDescription);
exports[userDescription.name] = addResource(userDescription);
exports[commentDescription.name] = addResource(commentDescription);
exports[videoDescription.name] = addResource(videoDescription);
exports[layoutDescription.name] = addResource(layoutDescription);
exports[datasetDescription.name] = addResource(datasetDescription);


exports['helpers'] = {
    /** Takes care of the details of handling a panel and a dataset.
     *
     * @param panel The panel resource (red9panel)
     * @param resource The require'd resource (red9resource)
     * @param newDataset The dataset object to store in the DB
     * @param panelStream A stream representing the panel data
     * @param callback {err, updatedDataset}
     * @param bool never set to true. Leave undefined.
     */
    createDataset: function createDataset(panel, resource, newDataset, panelStream, callback, deepMigrate) {
        resource.dataset.create(newDataset, function (err, createdDataset) {
            if (err) {
                console.log(err);
            }
            panel.createPanel(createdDataset.id, panelStream, function (err) {
                if (err) {
                    callback(err);
                } else {
                    setTimeout(function () {
                        panel.readPanelJSON(createdDataset.id, {
                            properties: {},
                            statistics: {},
                            csPeriod: 1000000
                        }, function (err, result) {
                            if (err) {
                                callback(err);
                            } else {
                                resource.dataset.update(createdDataset.id, {
                                    startTime: result.startTime,
                                    endTime: result.endTime,
                                    summaryStatistics: result.summaryStatistics
                                }, function (err, updatedDataset) {
                                    if (err) {
                                        callback(err);
                                    } else {
                                        callback(null, updatedDataset);
                                    }
                                });
                            }
                        });
                    }, 500);
                }
            });
        }, deepMigrate);
    }
};
