// red9resource/lib/resource.description.layout.js
var underscore = require('underscore')._;
var Joi = require('joi');
var validators = require('./validators');

var basicModel = {
    id: validators.id,
    title: Joi.string().description('A short title for this layout'),
    description: Joi.string().description('Human readable description of this layout'),
    for: Joi.array().includes(Joi.string()).description('Routes that this layout is applicable for'),
    layout: Joi.any().description('The actual layout description')
};

var resourceName = 'layout';


module.exports = {
    name: resourceName,
    tableName: 'layout',

    models: {
        model: basicModel,
        create: Joi.object({
            title: basicModel.title.required(),
            description: basicModel.description.required(),
            for: basicModel.for.required(),
            layout: basicModel.layout.required()
        }).options({className: resourceName + '.create'}),
        update: Joi.object({
            title: basicModel.title,
            description: basicModel.description,
            for: basicModel.for,
            layout: basicModel.layout
        }).options({className: resourceName + '.update'}),
        resultModel: Joi.object({
            id: basicModel.id.required(),
            title: basicModel.title.required(),
            description: basicModel.description.required(),
            for: basicModel.for.required(),
            layout: basicModel.layout.required()
        }).options({className: resourceName}),
        search: {
            id: validators.multiCSV(basicModel.id),
        }
    },

    mapping: [
        {
            cassandraKey: 'id',
            cassandraType: 'uuid',
            jsKey: 'id',
            jsType: 'string'
        },
        {
            cassandraKey: 'title',
            cassandraType: 'varchar',
            jsKey: 'title',
            jsType: 'string'
        },
        {
            cassandraKey: 'description',
            cassandraType: 'varchar',
            jsKey: 'description',
            jsType: 'string'
        },
        {
            cassandraKey: 'for',
            cassandraType: 'varchar',
            jsKey: 'for',
            jsType: 'object'
        },
        {
            cassandraKey: 'layout',
            cassandraType: 'varchar',
            jsKey: 'layout',
            jsType: 'object'
        }
    ],

    checkResource: function (layout, callback) {
        callback(null, layout);
    }
};
