// red9resource/lib/resource.description.video.js
var moment = require('moment');
var _ = require('underscore')._;

module.exports = {
    name: 'video',
    tableName: 'video',

    mapping: [
        {
            cassandraKey: 'id',
            cassandraType: 'uuid',
            jsKey: 'id',
            jsType: 'string'
        },
        {
            cassandraKey: 'start_time',
            cassandraType: 'timestamp',
            jsKey: 'startTime',
            jsType: 'timestamp'
        },
        {
            cassandraKey: 'host',
            cassandraType: 'varchar',
            jsKey: 'host',
            jsType: 'string'
        },
        {
            cassandraKey: 'host_id',
            cassandraType: 'varchar',
            jsKey: 'hostId',
            jsType: 'string'
        },
        {
            cassandraKey: 'dataset',
            cassandraType: 'uuid',
            jsKey: 'dataset',
            jsType: 'string',
        },
        {
            cassandraKey: 'create_time',
            cassandraType: 'timestamp',
            jsKey: 'createTime',
            jsType: 'timestamp'

        }
    ],

    checkResource: function (video, callback) {
        // TODO(SRLM): Add checks:
        // - dataset must exist
        // - startTime < dataset endTime
        // - startTime > dataset startTime - N hours (allow for starting a little bit before
        // - host and hostID are valid (and exist and are public)

        callback(null);
    },
    populateOnCreate: function (newVideo) {
        newVideo.createTime = moment().valueOf();
    }
};
