// red9resource/lib/resource.crud.js
var moment = require('moment');
var common = require('./resource.common.js');
var cassandra = require('./cassandra');
var _ = require('underscore')._;
var VError = require('verror');

/**
 *
 * @param resourceDescription
 * @param newResource
 * @param callback {err, ???}
 */
exports.create = function (resourceDescription, newResource, callback) {
    if (_.isFunction(resourceDescription.populateDefaults)) {
        resourceDescription.populateDefaults(newResource);
    }
    // We're not going to check. Assume that all input is valid.

    // Every resource gets an ID
    newResource.id = common.generateUUID();
    if (_.isFunction(resourceDescription.populateOnCreate)) {
        resourceDescription.populateOnCreate(newResource);
    }

    resourceDescription.checkResource(newResource, function (err) {
        if (err) {
            callback(err);
            return;
        }

        var t = common.createResourceQuery(resourceDescription.tableName, resourceDescription.mapping, newResource);

        cassandra.execute({
            query: t.query,
            parameters: t.parameters,
            callback: function (err) {
                if (err !== null) {
                    callback(err);
                } else {
                    callback(null, newResource);
                }
            }
        });
    });
};

/**
 *
 * @param query
 * @param options
 * @param rowCallback
 * @param callback (err, totalRows)
 */
exports.find = function (resourceDescription, query, options, rowCallback, callback) {
    var dividedQuery;
    var pipeline;

    if (typeof options === 'undefined') {
        options = {};
    }
    if (!_.isFunction(rowCallback)) {
        rowCallback = function () {
        };
    }
    if (!_.isFunction(callback)) {
        throw new VError('must define a callback function');
    }
    dividedQuery = common.divideQueries(query);

    var cassandraQuery = common.constructWhereQueryWithParameters(resourceDescription.tableName,
        resourceDescription.mapping, dividedQuery.cassandra);

    pipeline = common.queryTailPipeline(options, resourceDescription.expand, rowCallback, callback);

    cassandra.execute({
        query: cassandraQuery.query,
        parameters: cassandraQuery.parameters,
        rowCallback: function (cassandraResource) {
            var resource = common.mapToResource(resourceDescription.mapping, cassandraResource);
            // var resource = resourceDescription.mapToResource(cassandraResource);
            if (_.isFunction(resourceDescription.populateDynamic)) {
                resource = resourceDescription.populateDynamic(resource);
            }

            if (common.testAgainstQuery(resource, dividedQuery.local)) {
                pipeline.row(resource);
            }
        },
        callback: function (err) {
            if (err) {
                callback(err);
            } else {
                // At this point the pipeline will call the done callback
                pipeline.done();
            }
        }

    });

};

/**
 *
 * @param find
 * @param resourceDescription
 * @param id
 * @param updatedResource
 * @param callback
 */
exports.update = function (find, resourceDescription, id, updatedResource, callback) {
    delete updatedResource.id; // if it has it...

    // Test to make sure there's something to update. If there isn't, then fail fast.
    var updateQuery = common.createUpdateQuery(resourceDescription.tableName, id, resourceDescription.mapping, updatedResource);
    if (updateQuery.parameters.length === 0) {
        callback(new VError('Must have at least one value for creating an update string.'));
        return;
    }

    // Check to make sure that the event exists
    // And store it so we can validate the updated event
    var resourceTemp;
    find({id: id}, null,
        function (foundResource) {
            resourceTemp = foundResource;
        },
        function (err, rowCount) {
            if (err) {
                callback(err);
            } else if (rowCount !== 1) {
                callback(new VError('%s %s does not exist (%s results found)', resourceDescription.tableName, id, rowCount));
            } else {
                resourceDescription.checkResource(_.extend(resourceTemp, updatedResource), function (err) {
                    if (err) {
                        callback(err);
                    } else {
                        cassandra.execute({
                            query: updateQuery.query,
                            parameters: updateQuery.parameters,
                            //query: queryString,
                            callback: callback
                        });
                    }
                });
            }
        });
};

/**
 *
 * @param find
 * @param resourceDescription
 * @param id
 * @param callback
 */
exports.delete = function (find, resourceDescription, id, callback) {
    var resource;
    find({id: id}, null,
        function (t) {
            resource = t;
        },
        function (err, rowCount) {
            if (err) {
                callback(err);
            } else if (rowCount !== 1) {
                callback(new VError('%s %s does not exist (%s results found)', resourceDescription.tableName, id, rowCount));
            } else {
                cassandra.execute({
                    query: common.createDeleteString(resourceDescription.tableName, id),
                    callback: function (err) {
                        if (err) {
                            callback(err);
                        } else {
                            callback(null, resource);
                        }
                    }
                });
            }
        });
};
