// red9resource/lib/validators.js
"use strict";

var Joi = require('joi');

// This second regex (after the OR) is a legacy option!!! As it turns out, I haven't been using version 4
var idRegex = new RegExp(/([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}|[a-f0-9]{8}-[a-f0-9]{4}-4[a-f0-9]{3}-[89aAbB][a-f0-9]{3}-[a-f0-9]{12})/);
var singleId = new RegExp('^' + idRegex.source + '$');
var csvId = new RegExp('^(' + idRegex.source + ',)*' + idRegex.source + '$');

exports.id = Joi.string().regex(singleId).description('unique Red9 UUID key');
exports.idCSV = Joi.string().regex(csvId).description('multiple unique Red9 UUID keys in a CSV list.');
exports.timestamp = Joi.number().integer().min(0).unit('milliseconds').description('milliseconds since epoch (Unix time)');
exports.duration = Joi.number().integer().min(1).unit('milliseconds');
exports.multiArray = function (single) {
    return Joi.alternatives(single, Joi.array().includes(single)).description('multiple key instances or array of ' + single._description);
};

exports.summaryStatistics = Joi.object().description('numerically summarizes a period of time from a panel').options({
    className: 'summaryStatistics'
});
